pennylane/gradients/vjp.py

```diff
@@ -15,11 +15,186 @@
+    return jac
+
+
+def compute_vjp_single(dy, jac):
+    """Convenience function to compute the vector-Jacobian product for a given
+    vector of gradient outputs and a Jacobian for a single measurement tape.
+
+    Args:
+        dy (tensor_like): vector of gradient outputs
+        jac (tensor_like, tuple): Jacobian matrix
+
+    Returns:
+        tensor_like: the vector-Jacobian product
+
+    **Examples**
+
+    1. For a single parameter and a single measurement without shape:
+
+    .. code-block:: pycon
+
+        >>> jac = np.array(0.1)
+        >>> dy = np.array(2)
+        >>> compute_vjp_single(dy, jac)
+        np.array(0.2)
+
+    2. For a single parameter and a single measurment with shape:
+
+    .. code-block:: pycon
+
+        >>> jac = np.array([0.1, 0.2])
+        >>> dy = np.array([1.0, 1.0])
+        >>> compute_vjp_single(dy, jac)
+        np.array([0.3])
+
+
+    3. For multiple parameters (in this case 2 parameters) and a single measurement without shape:
+
+    .. code-block:: pycon
+
+        >>> jac = tuple([np.array(0.1), np.array(0.2)])
+        >>> dy = np.array(2)
+        >>> compute_vjp_single(dy, jac)
+        np.array([0.2, 0.4])
+
+    3. For multiple parameters (in this case 2 parameters) and a single measurement with shape:
+
+    .. code-block:: pycon
+
+        >>> jac = tuple([np.array([0.1, 0.2]), np.array([0.3, 0.4])])
+        >>> dy = np.array([1.0, 2.0])
+        >>> compute_vjp_single(dy, jac)
+        np.array([0.5, 1.1])
+
+    """
+    if jac is None:
+        return None
+
+    dy_row = math.reshape(dy, [-1])
+
+    if not isinstance(dy_row, np.ndarray):
+        jac = _convert(jac, dy_row)
+
+    try:
+        if math.allclose(dy, 0):
+            # If the dy vector is zero, then the
+            # corresponding element of the VJP will be zero.
+            if isinstance(jac, tuple):
+                num_params = len(jac)
+            else:
+                num_params = 1
+            res = math.convert_like(np.zeros([num_params]), dy)
+            return math.cast(res, dy.dtype)
+    except (AttributeError, TypeError):
+        pass
+
+    # Single measurement with a single param
+    if not isinstance(jac, (tuple, autograd.builtins.SequenceBox)):
+        # No trainable parameters
+        if jac.shape == (0,):
+            res = qml.math.zeros((1, 0))
+            return res
+        # Single measurement with no dimension e.g. expval
+        if dy.shape == ():
+            jac = math.reshape(qml.math.squeeze(jac), (1, 1))
+            res = math.tensordot(jac, dy_row, [[0], [0]])
+        # Single measurement with dimension e.g. probs
+        else:
+            jac = math.reshape(jac, (len(jac), 1))
+            res = math.tensordot(jac, dy_row, [[0], [0]])
+    # Single measurement with multiple params
+    else:
+        # No trainable parameters (adjoint)
+        if len(jac) == 0:
+            res = qml.math.zeros((1, 0))
+            return res

```

Comment:
The second case is there because jac is a tuple and therefore does not have a shape attribute.

Reply:
Right, but `len(jac) == 0` should work both in the case of `jac` being a tuple or array, right? I.e., the first check could be substituted with `len(jac) == 0` instead of `if jac.shape == (0,)` and then the two cases are identical.

This is minor though overall.

Reply:
When the shape is (0,) the length is not defined.